### PR TITLE
wal: never leave a corrupted wal file

### PR DESCRIPTION
**Pull request by xiang90**
If the process dies during wal.cut(), it might leave a corrupted wal
file. This commit solves the problem by creating a temp wal file first,
then atomically rename it to a wal file when we are sure it is vaild.

### CODE REVIEW

File: wal/wal.go [hunk 1]
```diff
@@ -284,31 +284,28 @@ func (w *WAL) ReadAll() (metadata []byte, state raftpb.HardState, ents []raftpb.
...
+	if err := w.f.Close(); err != nil {
 		return err
 	}
-	err = l.Lock()
+
+	fpath := path.Join(w.dir, walName(w.seq+1, w.enti+1))
+	ftpath := fpath + ".tmp"
+
+	// create a temp wal file with name sequence + 1, or tuncate the existing one
+	ft, err := os.OpenFile(ftpath, os.O_WRONLY|os.O_APPEND|os.O_CREATE|os.O_TRUNC, 0600)
```
Comment (yichengq) [inline wal/wal.go]: we dont' need O_APPEND any more, right?

Comment (xiang90) [inline wal/wal.go]: we need for safety. so you can never seek back.

**Comment by yichengq** [comment]
lgtm

**Pull request merged**